asset: Add Asset versioning support - V1

jharveyb opened the pull request:
Fixes https://github.com/lightninglabs/taproot-assets/issues/392.

jharveyb commented:
This PR is now part 1 of 3ish for actually bumping the asset version.

This includes the new fields in the `Asset` struct, without being populated, new encoders/decoders, the new TLV numbers + record funcs, and rejection of unknown asset versions before they are used in an MS-SMT.

One piece missing is rejection of unknown assets for RPC inputs.

Roasbeef commented on asset/asset.go:
```diff
@@ -22,12 +23,20 @@ import (
 
+// SchnorrKey is a type for representing an x-only public key, serialized in the
+// 32-byte form.
+type SchnorrKey [32]byte
```
👍

Roasbeef commented on asset/asset.go:
```diff
@@ -22,12 +23,20 @@ import (
 	return s[1:]
 }
 
+func (s SchnorrKey) SchnorrSerialized() []byte {
```
Should we start to make this into a slim interface? Or are the type defs above sufficient?

jharveyb replied:
Not sure what you mean wrt. slim interface - I think using the `any` type would cause trouble elsewhere. Maybe we can do something with a custom type constraint though?

Roasbeef replied:
Mean moreso making `SchnorrSerialized` into an interface and using that elsewhere. Then either key can be passed in depending on context. I think we want to have this be explicit though, so we can ensure we're using the correct serialization when necessary.

Roasbeef commented on asset/asset.go:
```diff
@@ -36,6 +45,15 @@ func (s SerializedKey) CopyBytes() []byte {
+	return fn.CopyBytes(s[:])
+}
+
+func (s SchnorrKey) IsZero() bool {
```
Missing `godoc` comment.

Roasbeef commented on asset/asset.go:
```diff
@@ -53,6 +79,8 @@ var (
 	// asset split leaves.
 	ZeroPrevID PrevID
 
+	ZeroGroupKey [32]byte
```
Missing `godoc` comment. Also looks like we can use `SchnorrKey` here.

Roasbeef commented on asset/asset.go:
```diff
@@ -658,6 +686,14 @@ type Asset struct {
 	Genesis
 
+	// Type uniquely identifies the type of Taproot asset.
+	ExplicitType Type
```
Why's this needed? 

I would imagine that we can actually keep the `Genesis` struct here as is. The difference would be that upon decode, you would then "supplement" this information from the DB or elsewhere. In most cases, you still want to know this information, but you might obtain it in an extra step. The universe changes made a similar gap where we already know the genesis before decoding the asset, so we supplement that as needed.

jharveyb replied:
AH true, I forgot about that part of offline discussion.

I think the group key field would still change though, since the sig is being moved into the witness.

Roasbeef replied:
Yeah true, the key changes as now it's _just_ a key and not also the sig that comes along side it.

Roasbeef commented on asset/asset.go:
```diff
@@ -692,6 +728,11 @@ type Asset struct {
+	// XOnlyGroupKey is the tweaked public key that is used to associate
+	// assets together across distinct asset IDs, allowing further issuance
+	// of the asset to be made possible.
+	XOnlyGroupKey SchnorrKey
```
Similar here as well: for a logical asset we want to know this information, but it can be obtained via another source.

jharveyb replied:
So maybe we actually want to modify the `AssetGroup` struct to include the tapscript root for the group key? For it to maintain the same definition of group key + info needed to derive it.

Roasbeef replied:
Wouldn't that just live in the initial genesis proof? At that point you've already verified the derivation. 

I do agree that we need to store the group key derivation explicitly, so this way we can refer back to it (verify the inclusion proof for the tapscript leaf, etc) when it comes to further minting.

Roasbeef replied:
> So maybe we actually want to modify the AssetGroup struct to include the tapscript root for the group key?

Or in something like `GroupDerivationInfo` or w/e which is read when we need to go verify subsequent minting events?

Roasbeef commented on asset/records.go:
```diff
@@ -31,6 +26,15 @@ const (
+	// TLV itself as done with LeafGenesis.
+	LeafAssetID       LeafTlvType = 2
+	LeafAssetType     LeafTlvType = 4
+	LeafXOnlyGroupKey LeafTlvType = 17
```
17? 

Why also why would ID and type need to be serialized? You only need the asset ID. You already have the genesis after initial boostrap, so from that you can verify that the asset ID here is valid.

Roasbeef commented on asset/records.go:
```diff
@@ -49,8 +53,11 @@ func NewLeafVersionRecord(version *Version) tlv.Record {
-	return tlv.MakePrimitiveRecord(LeafID, id)
+func NewLeafIDRecord(id *ID) tlv.Record {
+	const recordSize = sha256.Size
+	return tlv.MakeStaticRecord(LeafAssetID, id, recordSize, IDEncoder,
```
Reason this changed was just to be able to use this new concrete type?

Roasbeef commented on asset/records.go:
```diff
@@ -143,6 +150,13 @@ func NewLeafScriptKeyRecord(scriptKey **btcec.PublicKey) tlv.Record {
 	)
 }
 
+func NewLeafXOnlyGroupKeyRecord(groupKey **GroupKey) tlv.Record {
```
See above, not sure we'll need this.

jharveyb replied:
I think we do want it? If we're leaving in the current logic for group key struct encode/decode in this PR.

Changing it now to pull only the key from `GroupKey` though instead of using a new field in the `Asset` struct.

Roasbeef replied:
> I think we do want to change the encoding to only the x-only key though? Just not change the top-level struct.

x-only for the group key? Rationale being that it's now an actual taproot output key so it should itself be 32-bytes I take it?

jharveyb replied:
Yeah, right now we actually don't follow the BIP since we do serde for the group key as compressed 33-byte vs. Schnorr 32-byte.

https://github.com/lightninglabs/taproot-assets/blob/429b3ea2762ba5ef8a351d53454c672c2a54017b/asset/encoding.go#L557

Compared to the script key,  this key will always be 32-byte b/c it can only be derived with a TapTweak, so we could just do serde with 32 bytes instead of 33.

A separate question is whether we should store it in the DB as 32 or 33 bytes - it's 33 right now, IIRC that was for consistency with all the other keys we store?

Roasbeef commented on fn/memory.go:
```diff
@@ -36,3 +36,10 @@ func CopySlice[T any](slice []T) []T {
 }
+
+// CopyBytes returns a deep copy of the given byte slice.
+func CopyBytes(b []byte) []byte {
```
Isn't htis the same as `CopySlice[[]byte]`?

Roasbeef replied:
https://go.dev/play/p/_D5C9txX2TI

jharveyb replied:
Yup, I misinterpreted the 'shallow copy' warning on `CopySlice`.

Roasbeef commented on tapdb/sqlc/migrations/000002_assets.up.sql:
```diff
@@ -172,6 +172,10 @@ CREATE TABLE IF NOT EXISTS assets (
     
     genesis_id INTEGER NOT NULL REFERENCES genesis_assets(gen_asset_id),
 
+    asset_computed_id BLOB UNIQUE,
```
Same here, I don't think we need any of these fields. We'll already know the genesis information before we even are sent an asset. That information will always be there, as it must be known in order to properly verify an asset. 

In other words, we shouldn't really need to change this on disk structure. What we're changing is the wire level serialization: where these fields are encoded vs not. Ideally all the higher level logic stays the exact same (or needs minimal changes at least).

jharveyb replied:
I think we still need changes for the group key side though, to now store the tapscript root?

jharveyb replied:
Specifically a new field in the `asset_groups` table to store tapscript roots.

Roasbeef replied:
> Specifically a new field in the asset_groups table to store tapscript roots.

Yep agreed this is necessary. I think the other question this leads into is: shoudl we break out the group key derivation info into another sub-table? Today we have the ability to crate and spend form script keys that have script paths, this logic is done primary outside the daemon though. Given that minting right now is something fully encapsulated in the daemon, do we also need to store the leaves as well for the group key tapscript tree, or would we rely again on the caller to handle that?

jharveyb replied:
Discussed offline, resolved to just store the root for now.

Roasbeef commented on asset/asset.go:
```diff
@@ -735,6 +740,12 @@ type Asset struct {
+// IsUnknownVersion returns true if an asset has a version that is not
+// recognized by this implementation of tap.
+func (a *Asset) IsUnknownVersion() bool {
+	return a.Version != V0
```
Morph into a switch statement now?

Roasbeef commented on proof/proof.go:
```diff
@@ -99,6 +99,16 @@ type Proof struct {
+
+	// GenesisReveal is an optional Genesis struct, that must be provided
+	// for issuance proofs. This allows for derivation of the asset ID.
+	GenesisReveal *asset.Genesis
```
👍

Roasbeef commented on proof/proof.go:
```diff
@@ -99,6 +99,16 @@ type Proof struct {
+	// and Tapscript root used to derive the group key for the asset group.
+	// The field is defined as internal_key || tapscript_root, and must be
+	// provided for issuance proofs of grouped assets.
+	GroupKeyReveal *asset.GroupKeyReveal
```
Add additional context here (and also the `GroupKeyReveal` struct) on the full derivation? So the first tweak with the asset ID info, then the extra layer with the actual tapscript root.

jharveyb replied:
Added as godoc for the actual `GroupKeyReveal` struct.

guggero commented on asset/asset.go:
```diff
@@ -36,6 +45,15 @@ func (s SerializedKey) CopyBytes() []byte {
+}
+
+func (s SchnorrKey) IsZero() bool {
+	return bytes.Equal(s[:], ZeroGroupKey[:])
```
Since they're both of the same type and an array, can just use `s == ZeroGroupKey` (which I imagine is faster as well).

guggero commented on asset/asset.go:
```diff
@@ -658,6 +686,14 @@ type Asset struct {
+	// Type uniquely identifies the type of Taproot asset.
+	ExplicitType Type
+
+	// ID is an asset's unique identifier within the Taproot Asset protocol.
```
nit: Godoc comment name doesn't match struct member name (here and for `ExplicitType`).

guggero reviewed (commented):
Did a quick first pass to grok the changes. From Laolu's comments it sounds like this might change quite a bit soon, so will do a second pass after those changes.

halseth commented on asset/asset.go:
```diff
@@ -28,6 +28,11 @@ type SerializedKey [33]byte
+// GroupKeyReveal is a type for representing the data used to derive a group key
+// for an asset. Specifically, internal_key || tapscript_root.
+// TODO(jhb): Implement constructor for this field that includes length checks
+type GroupKeyReveal [GroupKeyRevealLen]byte
```
Would it make more sense to make this a struct 
```go
type GroupKeyReveal {
    Key [32]byte
    Root [32]byte
}
```
that we could deserialize directly into?

jharveyb commented:
Removed the DB changes for assets, and made `GroupKeyReveal` a struct vs. flat byte array. Updated TLV numbers and reverted the changes to the asset struct.

IIUC the only DB change needed is a new field to store the actual group key tapscript root? @Roasbeef

Roasbeef commented:
> IIUC the only DB change needed is a new field to store the actual group key tapscript root? @Roasbeef

Yep, I think so. Also see this comment on if we want to also store the script leaf information in the DB as well, if we assume the caller would do that: https://github.com/lightninglabs/taproot-assets/pull/381#discussion_r1279988255. If the latter, then we may need to expose some more utility RPCs to allow the caller to create a valid minting witness for their chosen part (so the give us the control block, etc).

jharveyb commented:
Following offline discussion, the current plan is:

 - Keep the group key as 33-byte, since `TapTweak` won't guarantee even parity for the final tweaked key. This also simplifies key handling in the DB overall.

(This requires a BIP update).

 - Adapt the asset DB tables to move away from storing the minting witness in the `asset_group_sigs` table, and instead store in the `asset_witness` table or a new table like `group_witnesses`. We should end up with no duplication of the minting witnesses, and queries to fetch these witnesses shouldn't be a large change.

halseth commented on asset/asset.go:
```diff
@@ -459,6 +463,21 @@ type GroupKey struct {
+type GroupKeyReveal struct {
+	// InternalTweakedKey is the tweaked public key that is a result of:
+	// groupInternalKey + sha256(groupInternalKey || assetID) * G.
+	InternalTweakedKey SchnorrKey
```
This should be the non-tweaked key? (called `asset_raw_group_key`) in the BIP. Otherwise one cannot it has been tweaked with the `asset_id`

jharveyb replied:
Ah true, I think I was mixing this up with `GroupKey.GroupPubKey`, which would now be the doubly-tweaked key.

jharveyb commented:
Ok, this has been updated to meet the discussed changes:

 - Tweaked group key is back to 33 bytes.
 - Better godoc comments explaining the tweaking process.
 - New DB table to store witnesses for asset group reissuance, that stores a tapscript root and an arbitrary witness stack.
 - Updates to other DB logic to maintain current functionality.
 
 AFAICT, those DB updates went well (unit tests on SQLite passing), but there is some mistake in the change to the asset group info view. The new `UpsertAssetGroupWitness` query should also be better tested.
 
 The error caused by the view update:
 
 ```bash
 CREATE VIEW key_group_info_view AS
SELECT
sig_id, gen_asset_id, tapscript_root, witness_stack, tweaked_group_key,
        raw_key, key_index, key_family
FROM asset_group_sigs sigs
JOIN asset_group_witnesses witnesses
        ON sigs.witness_id = witnesses.witness_id
JOIN asset_groups groups
        ON sigs.group_key_id = groups.group_id
JOIN internal_keys keys
        ON keys.key_id = groups.internal_key_id
WHERE sigs.gen_asset_id IN (SELECT gen_asset_id FROM genesis_info_view);
(details: ERROR: relation "asset_group_witnesses" does not exist (SQLSTATE 42P01))
Test:       	TestSelectCommitment/asset_with_key_group
 ```

guggero commented on asset/asset_test.go:
```diff
@@ -198,18 +198,20 @@ func TestAssetEncoding(t *testing.T) {
-			MetaHash:    [MetaHashLen]byte{1, 2, 3},
-			OutputIndex: 1,
-			Type:        1,
+	splitGen := Genesis{
```
nit (commit message): what does "serde" mean? Sounds like a vim command :)

guggero replied:
aaaah, okay, got it. how about "ser/de-ser"? Still pretty short but IMO a bit easier to decipher.

halseth replied:
I also thought it was some fancy unix command or something 😂

guggero commented on asset/records.go:
```diff
@@ -49,8 +51,11 @@ func NewLeafVersionRecord(version *Version) tlv.Record {
-	return tlv.MakePrimitiveRecord(LeafID, id)
+func NewLeafIDRecord(id *ID) tlv.Record {
+	const recordSize = sha256.Size
+	return tlv.MakeStaticRecord(LeafAssetID, id, recordSize, IDEncoder,
```
nit: formatting of multi-line function call.

guggero commented on proof/encoding.go:
```diff
@@ -336,3 +338,43 @@ func MetaTypeDecoder(r io.Reader, val any, buf *[8]byte, l uint64) error {
 	return tlv.NewTypeForEncodingErr(val, "MetaType")
 }
+
+func GroupKeyRevealEncoder(w io.Writer, val any, buf *[8]byte) error {
```
A unit test for these encoders/decoders would be nice.

guggero commented on tapdb/sqlc/migrations/000002_assets.up.sql:
```diff
@@ -107,14 +107,24 @@ CREATE TABLE IF NOT EXISTS asset_groups (
+CREATE TABLE IF NOT EXISTS asset_group_witnesses (
+    witness_id INTEGER PRIMARY KEY,
+
+    tapscript_root BLOB,
```
Do we hope that we can re-use some of these witnesses? Otherwise the alternative would be to just add both of these fields to the `asset_group_sigs` table and reduce the number of tables needed.

jharveyb replied:
I don't think so - I think actually the `tapscript_root` should go into `asset_groups` since it is static per group key, and then `witness_stack` could indeed be a field in `asset_group_sigs`.

One consideration is that the witnesses aren't really bounded size right now, and I suppose if they are in a separate table then we have more options in not fetching them for some queries? Not sure if that's a relevant concern.

halseth replied:
Agree that it would make sense to have the static `tapscript_root` with the asset group itself. No strong opinion on where the witness stack goes.

guggero commented on tapdb/sqlc/queries/assets.sql:
```diff
@@ -155,13 +155,21 @@ RETURNING group_id;
 RETURNING sig_id;
 
+-- name: UpsertAssetGroupWitness :one
+INSERT INTO asset_group_witnesses (
```
This isn't really an upsert, as there is nothing to conflict on here. So it will always insert a new record...

guggero commented on asset/asset.go:
```diff
@@ -488,6 +488,85 @@ func (g *GroupKey) IsEqualGroup(otherGroupKey *GroupKey) bool {
+
+	// We can safely drop the signature parsing error, since a witness that
+	// isn't a signature may still be a valid script witness.
+	if len(potentialSig) == schnorr.SignatureSize {
```
nit: invert condition and continue the happy path non-indented?

guggero commented on asset/asset.go:
```diff
@@ -488,6 +488,85 @@ func (g *GroupKey) IsEqualGroup(otherGroupKey *GroupKey) bool {
+
+	witnessStack = make([]byte, b.Len())
+	copy(witnessStack, b.Bytes())
+	return witnessStack, nil
```
I think `b.Bytes()` of a `bytes.Buffer` is safe to use without an additional copy.

guggero commented on asset/asset.go:
```diff
@@ -488,6 +488,85 @@ func (g *GroupKey) IsEqualGroup(otherGroupKey *GroupKey) bool {
+
+	err := TxWitnessEncoder(&b, &witness, &buf)
+	if err != nil {
+		return nil, fmt.Errorf(
```
nit: formatting of `Errorf()` call can use the compact style. Here and below.

guggero commented on asset/asset.go:
```diff
@@ -488,6 +488,85 @@ func (g *GroupKey) IsEqualGroup(otherGroupKey *GroupKey) bool {
+		return groupSig
+	}
+
+	return nil
```
Maybe have a second boolean parameter instead of returning nil for an `Is...()` method?

guggero reviewed (commented):
Looks pretty good, nice work!

halseth commented on asset/asset.go:
```diff
@@ -427,6 +431,23 @@ type GroupKey struct {
+type GroupKeyReveal struct {
+	// RawGroupKey is the public key that is tweaked twice to derive the
+	// final tweaked group key. The final tweaked key is the result of:
+	// InternalKey =  RawGroupKey * sha256(InternalKey || assetID) * G.
```
typo:  `InternalKey =  RawGroupKey * sha256(RawGroupKey || assetID) * G`

halseth commented on asset/asset.go:
```diff
@@ -427,6 +431,23 @@ type GroupKey struct {
 
+// GroupKeyReveal is a type for representing the data used to derive the tweaked
+// key used to identify an asset group. The final tweaked key is the result of:
+// TapTweak(internalTweakedKey, tapscriptRoot)
```
nit: inconsistent with naming below

jharveyb replied:
Aligned names to those used in the godoc for the `GroupKey` struct.

halseth commented on proof/proof.go:
```diff
@@ -99,6 +99,18 @@ type Proof struct {
+	// GenesisReveal is the Genesis information for an asset, that must be
+	// provided for issuance proofs, and must be empty for non-issuance
+	// proofs. This allows for derivation of the asset ID and rederivation
+	// of an asset group key.
```
nit: maybe mention that not all assets will have an asset group key.

halseth commented on proof/file.go:
```diff
@@ -190,11 +194,35 @@ func (f *File) Decode(r io.Reader) error {
 
+// IsUnknownVersion returns true if a proof has a version that is not
+// recognized by this implementation of tap.
+func (f *File) IsUnknownVersion() bool {
```
same nit as earlier

halseth commented on asset/asset.go:
```diff
@@ -488,6 +488,85 @@ func (g *GroupKey) IsEqualGroup(otherGroupKey *GroupKey) bool {
+		potentialSig = witness[0]
+	}
+
+	// We can safely drop the signature parsing error, since a witness that
```
This comment seems inconsistent with the method godoc?

halseth reviewed (changes requested):
Looks very good 😄

jharveyb commented on asset/asset.go:
```diff
@@ -694,6 +699,17 @@ type Asset struct {
 	GroupKey *GroupKey
 }
 
+// IsUnknownVersion returns true if an asset has a version that is not
```
I think `IsUnknown` works better, since we use that as the condition to fail early where this function is called.

Otherwise we end up with something like `if !asset.IsKnownVersion() {` and inverting the result anyways.

jharveyb commented:
Ok, I renamed `asset_group_sigs.genesis_sig` to `witness_stack` and added a `tapscript_root` field to `asset_groups`. The length for `tapscript_root` is not enforced at the DB level, but `GroupKeyReveal.TapscriptRoot` is `[32]byte` and not `[]byte` so I think that's fine for now.

IMO we can defer other DB field renaming once we actually add support for non-sig witnesses in another PR.

I added test coverage for `UpsertAssetGroupKey` and the helpers that serialize and parse witnesses. I also updated the proof encode/decode to fix some pointer issues.

Right now we only insert an empty tapscript root, since we only support single signatures as witnesses. That functionality should be deferred to a follow-up PR.

Roasbeef commented on proof/file.go:
```diff
@@ -203,8 +231,8 @@ func (f *File) NumProofs() int {
 // ProofAt returns the proof at the given index. If the file is empty, this
 // returns nil.
 func (f *File) ProofAt(index uint32) (*Proof, error) {
-	if f.IsEmpty() {
-		return nil, ErrNoProofAvailable
+	if err := f.IsValid(); err != nil {
```
I think we also want a version in the proof itself? https://github.com/lightninglabs/taproot-assets/blob/main/proof/proof.go#L43

So one top level is a version that describes how to parse and the serialization, etc. The other is a version for each proof that maybe tells how to validate, or which fields are expected to be there, etc.

guggero replied:
Added that as a last commit.

Roasbeef commented on tapdb/sqlc/migrations/000002_assets.up.sql:
```diff
@@ -107,7 +109,9 @@ CREATE TABLE IF NOT EXISTS asset_groups (
-    genesis_sig BLOB NOT NULL, 
+    -- The witness stack can contain either a single Schnorr signature for key
+    -- spends of the tweaked group key, or a more complex script witness.
+    witness_stack BLOB NOT NULL,
```
I think let's go ahead and rename now? Don't see much of a benefit of doing it later. Particularly since we'll merge this into a side branch that doesn't need to be fully working until we go to merge things back into master.

jharveyb replied:
Yup fair enough.

guggero replied:
We assumed that with the rename you meant also renaming the table from `asset_group_sigs` to `asset_group_witnesses`? That's addressed in the latest push.

Roasbeef commented on tapdb/sqlc/migrations/000002_assets.up.sql:
```diff
@@ -92,6 +92,8 @@ CREATE TABLE IF NOT EXISTS asset_groups (
 
     tweaked_group_key BLOB UNIQUE NOT NULL CHECK(length(tweaked_group_key) = 33), 
 
+    tapscript_root BLOB,
```
👍

Roasbeef commented on asset/asset.go:
```diff
@@ -487,6 +487,79 @@ func (g *GroupKey) IsEqualGroup(otherGroupKey *GroupKey) bool {
+// IsGroupSig attempts to parse the given witness as a single Schnorr signature,
+// and returns the signature and true if it is valid.
+func IsGroupSig(witness wire.TxWitness) (*schnorr.Signature, bool) {
+	if len(witness) != 1 {
```
These functions also need to mind the annex if we're going for 1:1 equivalence. There's some functions in `btcd/txscript` that should be useful for that.

guggero replied:
Or see https://github.com/lightninglabs/pool/blob/master/poolscript/script.go#L466

jharveyb replied:
Totally forgot about the annex, thanks for the catch - So IIUC for the current sig-only functionality, we should now allow any annex as long the first element is a sig?

Roasbeef commented on tapdb/assets_common.go:
```diff
@@ -259,8 +258,11 @@ func upsertGroupKey(ctx context.Context, groupKey *asset.GroupKey,
+	// support signatures as asset group key witnesses for now.
 	groupID, err := q.UpsertAssetGroupKey(ctx, AssetGroupKey{
 		TweakedGroupKey: tweakedKeyBytes,
+		TapscriptRoot:   []byte{},
```
Perhaps add this onto the struct now so harder to forget to insert it down the line? May save some debugging saga.

Roasbeef commented on tapdb/assets_common.go:
```diff
@@ -494,7 +502,8 @@ func parseGroupKeyInfo(tweakedKey, rawKey, genesisSig []byte,
 
-	groupSig, err := schnorr.ParseSignature(genesisSig)
+	// TODO(jhb): make generic for non-sig witnesses
+	groupSig, err := asset.ParseGroupSig(witness)
```
We'd then modify `GroupKey` to include both the reveal information and raw witness, no? Then that way given that you can verify the minting if needed. May depend on the actual call site though.

halseth replied:
When building on top of this PR I found at least the reveal information to be useful to be contained in the `GroupKey` struct.

jharveyb replied:
True, I guess it should sort of track the group key view. Will make them optional though, not sure if we have a call site where these would not be populated yet like importing the tweaked group key before having the witness or so.

Roasbeef reviewed (changes requested):
Super close. I think worth it to just do the rename now. We'll also want a version in the proof object itself. We should also handle the annex cases as rn we're still going for a 1:1 mapping.

halseth commented on asset/records.go:
```diff
@@ -31,6 +26,13 @@ const (
 	LeafGroupKey            LeafTlvType = 10
+
+	// Types for future asset format.
+	// TODO: LeafID and LeafGenesis have conflicting types. The BIP
```
this TODO is no longer relevant?

halseth commented on asset/asset.go:
```diff
@@ -427,6 +430,23 @@ type GroupKey struct {
+type GroupKeyReveal struct {
+	// RawGroupKey is the public key that is tweaked twice to derive the
+	// final tweaked group key. The final tweaked key is the result of:
+	// groupInternalKey =  RawKey * sha256(RawKey || assetID) * G.
```
BIP was updated to have the assetID be first in the sha256: https://github.com/Roasbeef/bips/pull/33#issuecomment-1669247706

guggero commented on tapdb/asset_minting.go:
```diff
@@ -572,7 +572,8 @@ func fetchAssetSprouts(ctx context.Context, q PendingAssetStore,
 			}
-			groupSig, err := schnorr.ParseSignature(sprout.GenesisSig)
+			// TODO(jhb): make generic for non-sig witnesses
+			groupSig, err := asset.ParseGroupSig(sprout.WitnessStack)
```
Maybe we should update the `asset.GroupKey` to contain the parsed `TxWitness` instead of the sig? Then we can move this TODO to where we validate the sig?

guggero reviewed (commented):
Getting very close indeed :tada:

jharveyb commented:
Ok, the PR has been updated to try and replace the group sig field with a witness, and add the witness and Tapscript root to the GroupKey struct. The problem as-is is that the nil-ness behavior for the sig as `*schnorr.Signature` and a similar use with `wire.TxWitness` / `[][]byte` are different.

Specifically, both fields are `BLOB NOT NULL` in the DB. While the sig is just converted to `[]byte` with Serialize(), we use the TLV encode/decode for witnesses here, which IIUC means that an uninitialized witness and an initialized but empty witness will serialize differently.

Separately from the encoding of a nil vs. empty witness, currently the parsing of a signature from a witness is currently possibly too strict, since an empty witness may not always be an error case. For example, in some of the Universe subsystem, we may load or store a GroupKey with the sig unset, simply because it isn't relevant at that point and/or was validated earlier.

One option for adding more consistency for storing the witness, is to actually have the field as just `BLOB`, and then check for both a nil or empty witness when parsing a stored witness instead of failing in case of a nil witness.

guggero commented:
I took this PR over while @jharveyb is absent. Spoke to @Roasbeef and the intention is to merge this to the `v0-3-0-staging` branch while not 100% working. For example the unit and integration tests will only work again once we actually add the group key signature as the asset witness in the minting proof.

But I think the PR is in a state where it is mostly complete for a first part and we can work towards fixing everything on this staging branch.

halseth commented on proof/proof_test.go:
```diff
@@ -505,6 +507,93 @@ func TestGenesisProofVerification(t *testing.T) {
+		// file index X", we add to the proof-file.hex file.
+		case strings.Contains(comment, "valid regtest proof file index"):
+			fileProofs = append(fileProofs, *proof)
+		}
```
add default case for unexpected comment?

guggero replied:
There are more than those three test cases in the file, we just export some of them into a different format to test. So we already ignore a bunch of them. Will make it more apparent by adding an empty default case with a comment.

halseth commented on asset/records.go:
```diff
@@ -15,12 +14,7 @@ import (
-	// preimage to the asset ID elsewhere, when it could be included in the
-	// TLV itself as done with LeafGenesis.
-	LeafID                  LeafTlvType = 1
+	LeafVersion             LeafTlvType = 0
```
`genesis` is replaced with pure `assetID` in the BIP, but we can make that change in a follow-up PR.

halseth commented on asset/asset.go:
```diff
@@ -422,6 +425,23 @@ type GroupKey struct {
+// key used to identify an asset group. The final tweaked key is the result of:
+// TapTweak(groupInternalKey, tapscriptRoot)
+type GroupKeyReveal struct {
+	// RawGroupKey is the public key that is tweaked twice to derive the
```
nit: godoc name mismatch

halseth commented on tapdb/assets_common.go:
```diff
@@ -275,10 +276,10 @@ func upsertGroupKey(ctx context.Context, groupKey *asset.GroupKey,
-		GenesisSig: groupKey.Sig.Serialize(),
-		GenAssetID: genAssetID,
-		GroupKeyID: groupID,
+	groupSigID, err := q.UpsertAssetGroupWitness(ctx, AssetGroupWitness{
```
nit: variable name still `sig`

halseth commented on asset/asset.go:
```diff
@@ -426,8 +427,17 @@ type GroupKey struct {
+	// script spend conditions for the group key. Instead of spending an
+	// asset, these scripts are used to define witnesses more complex than
+	// a Schnorr signature for reissuing assets.
+	TapscriptRoot [sha256.Size]byte
```
Important to note that this can be empty (key-spend only output key). Should we make this nillable, or is it good enough to have it be all zero?

guggero replied:
Good point. Turned it into a `[]byte` with a comment. Can do some additional validation later on.

halseth reviewed (approved):
Mostly just nits, will be good to have this merged in order to build on top 👍

halseth commented on proof/records.go:
```diff
@@ -286,3 +295,36 @@ func MetaRevealTypeRecord(metaType *MetaType) tlv.Record {
+func GroupKeyRevealRecord(reveal **asset.GroupKeyReveal) tlv.Record {
+	recordSize := func() uint64 {
+		if reveal == nil || *reveal == nil {
+			panic("nil reveal")
```
should we really panic here?

guggero commented:
Going to merge this into the `v0-3-0-staging` branch now so we can continue building on top. Can always fix remaining issues later on that branch.
